XFER-9166: Server and User resources

Alex-Vol-Amz opened the pull request:
Nearly 100% code coverage using unit tests.

*Issue XFER-9166*

*Description of changes:*

Added Server and User resources.

Both Server and User are passing 100% contract tests.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

lydakis commented on aws-transfer-user/contract-tests-artifacts/dependencies.yml:
```diff
@@ -0,0 +1,56 @@
+      Protocols:
+        - SFTP
+
+#  TestUser:
```
Do we want to keep this in?

Alex-Vol-Amz replied:
Oh, that was some unrelated test code. I should delete it.

Alex-Vol-Amz commented:
Final update after discussions with Uluru SME.

Changed the primaryIdentifier for Server and User to be the `Arn` matching the self-service schema.

Updated the handlers to manage the Arn correctly and extract the additionalIdentifiers from it when necessary.

Alex-Vol-Amz commented:
Deleted the tools that would only work internally and renamed the code named class to a generic name.

aws-sde reviewed (changes requested):
Still reviewing; will pick this up tomorrow but figured I'd publish the comments I have so far.

Alex-Vol-Amz commented on aws-transfer-server/src/main/java/software/amazon/transfer/server/BaseHandlerStd.java:
```diff
@@ -0,0 +1,361 @@
+    private ProgressEvent<ResourceModel, CallbackContext> handleRetryableServiceException(
+            ResourceModel model, CallbackContext ctx, String op, String reqToken, Exception ex) {
+
+        int currentNumThrottlingRetries = ctx.getNumThrottlingRetries();
+        if (currentNumThrottlingRetries > 0) {
+            ctx.setNumThrottlingRetries(currentNumThrottlingRetries - 1);
```
I suppose it is but it is simpler to think this way about the total retry like behaviors. I could drop the throttling qualifier is you prefer it not to be specific.

aws-sde replied:
Yeah I think it'd be nice to rename it but this is more of a nit

Alex-Vol-Amz commented on aws-transfer-server/src/main/java/software/amazon/transfer/server/BaseHandlerStd.java:
```diff
@@ -0,0 +1,361 @@
+    private ProgressEvent<ResourceModel, CallbackContext> handleThrottling(
+            ResourceModel model, CallbackContext ctx, String op, String reqToken, Exception ex) {
+
+        int currentNumThrottlingRetries = ctx.getNumThrottlingRetries();
+        if (currentNumThrottlingRetries > 0) {
+            ctx.setNumThrottlingRetries(currentNumThrottlingRetries - 1);
+            return ProgressEvent.defaultInProgressHandler(ctx, getDelaySeconds(ex), model);
+        }
```
All it means is that we should backoff and retry if necessary after a delay. I do not believe we should try and adjust the client retry policy. I prefer using the standard retry policies when possible and handle this here.

I picked this pattern from other handlers and I found it worth keeping.

aws-sde replied:
Why do we prefer to implement our own back off and retry mechanisms rather than relying on the SDK's built-in one?

Alex-Vol-Amz replied:
One reason is that we must respond to the calling state machine within 60 seconds or less. This mechanism is designed to use throttling based on standard retry strategy and then let the state machine do the overall retry so that it can remain in the IN_PROGRESS state. While we are retrying the rules for this chain reaction state machine dictate that we keep responding in a timely manner.

aws-sde replied:
Got it - responding within 60 seconds is a good point I was not aware of.

Alex-Vol-Amz commented on aws-transfer-server/src/main/java/software/amazon/transfer/server/BaseHandlerStd.java:
```diff
@@ -0,0 +1,361 @@
+    }
+
+    private void slowDownRetries(State state, String serverId) {
+        if (!Boolean.getBoolean("necco.uluru.unit.tests")) {
```
It is set in the pom.xml so we do not have dead time in unit tests. That way the test time is diminished without having to have sleep calls in the middle of testing a mock. I think I should make sure this is also set in the Brazil project files.

I will remove the code name.

Alex-Vol-Amz commented on aws-transfer-server/src/main/java/software/amazon/transfer/server/DeleteHandler.java:
```diff
@@ -0,0 +1,67 @@
+        try (TransferClient transferClient = client.client()) {
+            DeleteServerResponse awsResponse =
+                    client.injectCredentialsAndInvokeV2(awsRequest, transferClient::deleteServer);
+            logger.log(String.format("%s successfully deleted.", ResourceModel.TYPE_NAME));
```
Sure, I will adjust this.

aws-sde replied:
Thanks! Some of our current handler logs are missing the identifiers and I've been wishing they were there when log diving

Alex-Vol-Amz commented on aws-transfer-server/src/main/java/software/amazon/transfer/server/translators/ResourceModelAdapter.java:
```diff
@@ -0,0 +1,79 @@
+                                Translator.translateTagListToTagMap(
+                                        previousResourceModel.getTags()))
+                        .orElseGet(HashMap::new));
+        tagsMap.putAll(
+                Optional.ofNullable(request.getPreviousSystemTags()).orElseGet(HashMap::new));
```
These are already included. previousResourceModel.getTags() has them inside. We are just adding the system tags to the complete collection here.

aws-sde commented on aws-transfer-user/src/main/java/software/amazon/transfer/user/CreateHandler.java:
```diff
@@ -0,0 +1,125 @@
+        try (TransferClient transferClient = client.client()) {
+            CreateUserResponse response =
+                    client.injectCredentialsAndInvokeV2(request, transferClient::createUser);
+            log("user created successfully", request.userName());
```
Let's log the server ID as well (or the model's primaryIdentifier if that's available). Ditto for other user handler logs

Alex-Vol-Amz replied:
I can see if I should make the model available here to make it possible to log the Arn instead. I generate the Arn for a create user when the handler is called.

aws-sde commented on aws-transfer-user/src/main/java/software/amazon/transfer/user/CreateHandler.java:
```diff
@@ -0,0 +1,125 @@
+                                        callbackContext))
+                .then(
+                        progress ->
+                                removeTags(
```
How could there be any tags to remove when it's a new resource?

Alex-Vol-Amz replied:
Good point. Probably never triggers this code path.

aws-sde commented on aws-transfer-user/src/test/java/software/amazon/transfer/user/ListHandlerTest.java:
```diff
@@ -0,0 +1,70 @@
+        assertThat(response.getCallbackContext()).isNull();
+        assertThat(response.getCallbackDelaySeconds()).isEqualTo(0);
+        assertThat(response.getResourceModel()).isNull();
+        assertThat(response.getResourceModels()).isNotNull();
```
Can we make a stronger assertion here that the models are what we expected rather than just not null?

aws-sde commented on aws-transfer-user/src/test/java/software/amazon/transfer/user/UpdateHandlerTest.java:
```diff
@@ -0,0 +1,236 @@
+
+        updateAndAssertSuccess(current, desired);
+
+        current.setSshPublicKeys(Arrays.asList(key1, key2, key3));
+        desired.setSshPublicKeys(Collections.singletonList(key1));
```
Nit: can we use `List.of` or is this targetting Java 8?

Alex-Vol-Amz replied:
We are targeting Java 11 from what I recall. But I need to check if the target bytecode settings are reflecting this.

Alex-Vol-Amz commented on aws-transfer-server/src/test/java/software/amazon/transfer/server/ListHandlerTest.java:
```diff
@@ -0,0 +1,52 @@
+        assertThat(response.getMessage()).isNull();
+        assertThat(response.getErrorCode()).isNull();
+
+        verify(sdkClient, atLeastOnce()).listServers(any(ListServersRequest.class));
```
I can add more assertions to this and a few of the other simple tests.

Alex-Vol-Amz commented on aws-transfer-server/src/test/java/software/amazon/transfer/server/BaseHandlerStdTest.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    @Test
+    public void simpleHandleErrorLogicTest() {
```
I do not find value in exhaustive testing. Making every single permutation of a code path tested makes it much harder to refactor code in the future.

I can look into parameterized testing for some obvious choices.

aws-sde replied:
I'm not suggesting we add additional test cases or more coverage for this. Just that we split up the existing tests.

Here, there is a single test repeating the same assertions with different inputs (different exception types). If one assertion fails, the whole test fails and the other inputs don't get executed. I think if we make these inputs distinct test cases, then we can get better visibility into test failures and our tests will also be more readable.

aws-sde replied:
Or is JUnit smart enough to run all assertions before failing the test?
